### PR TITLE
PHP 8.0 support

### LINKED ISSUE #19: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [ ] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [ ] Modify `composer.json` to drop support for PHP less than 7.3
- [ ] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [ ] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [ ] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [ ] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by svycka**
closes #19

### CODE REVIEW

**Comment by svycka** [comment]
@boesing okay all green

File: composer.json [hunk 1]
```diff
@@ -25,17 +25,18 @@
         }
     },
     "require": {
-        "php": "^5.6 || ^7.0",
-        "laminas/laminas-stdlib": "^2.7.7 || ^3.1",
```
Comment (svycka) [inline composer.json:29]: can we drop 2.*?

File: composer.json [hunk 2]
```diff
@@ -25,17 +25,18 @@
...
+        "php": "^7.3 || ~8.0.0",
+        "laminas/laminas-stdlib": "^3.3",
         "laminas/laminas-zendframework-bridge": "^1.0"
     },
     "require-dev": {
         "laminas/laminas-coding-standard": "~1.0.0",
         "laminas/laminas-crypt": "^3.2.1",
-        "laminas/laminas-servicemanager": "^2.7.8 || ^3.3",
```
Comment (svycka) [inline composer.json:35]: can we drop 2.*?

Comment (boesing) [reply composer.json:35]: Yes, we can! 👍 
Just dont change the API of this project and we are good to go.

**Comment by svycka** [comment]
@boesing then I think it's ready for review

File: src/Encrypt/Openssl.php [hunk 3]
```diff
@@ -140,7 +143,10 @@ protected function _setKeys($keys)
                         throw new Exception\InvalidArgumentException("Private key '{$cert}' not valid");
                     }
 
-                    openssl_free_key($test);
+                    // PHP 8 automatically frees the key instance and deprecates the function
+                    if (PHP_VERSION_ID < 80000) {
+                        openssl_free_key($test);
+                    }
```
Comment (froschdesign) [inline src/Encrypt/Openssl.php]: This is a duplication of code logic and therefore it should be extracted as a private method.

File: .travis.yml [hunk 4]
```diff
@@ -6,44 +6,32 @@ cache:
...
+    - php: 7.4
       env:
         - DEPS=lowest
-    - php: 7.2
+    - php: 7.4
+      env:
+        - DEPS=latest
+    - php: 'nightly'
       env:
         - DEPS=latest
```
Comment (froschdesign) [inline .travis.yml]: Same as in all other components:
```suggestion
    - php: nightly
      env:
        - DEPS=lowest
    - php: nightly
      env:
        - DEPS=latest
```

File: src/Encrypt/Openssl.php [hunk 5]
```diff
@@ -464,4 +471,15 @@ public function toString()
     {
         return 'Openssl';
     }
+
+    /**
+     * Free key resource if necessary.
+     * PHP 8 automatically frees the key instance and deprecates the function
+     */
+    private function freeKeyResource($keys): void
```
Comment (froschdesign) [inline src/Encrypt/Openssl.php]: The DocBlock for `$keys` is missing.

File: src/Encrypt/Openssl.php [hunk 6]
```diff
@@ -364,11 +364,17 @@ public function encrypt($value)
...
-            openssl_free_key($key);
+        $crypt  = openssl_seal($value, $encrypted, $encryptedkeys, $keys, 'RC4');
+
+
+        // PHP 8 automatically frees the key instance and deprecates the function
+        if (PHP_VERSION_ID < 80000) {
+            foreach ($keys as $key) {
+                openssl_free_key($key);
+            }
         }
```
Comment (froschdesign) [inline src/Encrypt/Openssl.php]: The new method should also be used here.

File: src/Encrypt/Openssl.php (hunk 5 above)
Comment (svycka) [reply src/Encrypt/Openssl.php]: true :+1:

File: src/Encrypt/Openssl.php (hunk 6 above)
Comment (svycka) [reply src/Encrypt/Openssl.php]: I left it as micro optimization as it does not require to execute `foreach` block. Do you suggest moving this `foreach` to that method?

File: src/Encrypt/Openssl.php (hunk 5 above)
Comment (boesing) [reply src/Encrypt/Openssl.php]: Please add proper native typehint for `keys` here aswell as phpdoc to declare what keys are (I expect `array<int,string>` here.

File: src/Encrypt/Openssl.php [hunk 7]
```diff
@@ -364,11 +364,17 @@ public function encrypt($value)
             $value    = $compress($value);
         }
 
-        $crypt  = openssl_seal($value, $encrypted, $encryptedkeys, $keys);
-        foreach ($keys as $key) {
-            openssl_free_key($key);
```
Comment (boesing) [inline src/Encrypt/Openssl.php:369]: Please just change this function call to just target the new created method `Openssl:: freeKeyResource` to keep maintainability easy.

**Review by boesing — CHANGES_REQUESTED** [review]
Huge improvements here! Thanks!
I've added some change requests. If applied, we are ready to merge 👍

**Review by boesing — APPROVED** [review]
LGTM.

**Comment by boesing** [comment]
Trying to create a release next week.
Sorry for having you waiting...

File: .travis.yml [hunk 8]
```diff
@@ -12,38 +12,31 @@ env:
...
       env:
         - DEPS=lowest
-    - php: 7.1
+    - php: 7.4
       env:
         - DEPS=latest
-        - CS_CHECK=true
-        - TEST_COVERAGE=true
-    - php: 7.2
+    - php: 8.0
```
Comment (svycka) [inline .travis.yml:29]: php 8.0 available travis

File: .travis.yml [hunk 9]
```diff
@@ -12,38 +12,31 @@ env:
...
       env:
         - DEPS=lowest
-    - php: 7.2
+        - COMPOSER_ARGS="--no-interaction --ignore-platform-reqs"
```
Comment (svycka) [inline .travis.yml:32]: `--ignore-platform-reqs` required only for php8

**Review by boesing — APPROVED** [review]
LGTM

**Comment by boesing** [comment]
Thanks, @svycka!

**Pull request merged**